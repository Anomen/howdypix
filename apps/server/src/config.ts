// Load env variables
require("dotenv").config();

const config = {
  rabbitMq: {
    url: process.env.RABBITMQ_URL || "amqp://localhost"
  },
  serverApi: {
    baseUrl: "",
    port: parseInt(process.env.API_PORT as string) || 3004
  },
  auth: {
    code: {
      secret: "secret_code_string",
      expiry: "2h"
    },
    token: {
      secret: "secret_token_string",
      expiry: "4h"
    },
    refreshToken: {
      secret: "secret_refresh_string",
      expiry: "30d"
    }
  }
};

config.serverApi.baseUrl =
  process.env.API_BASE_URL || `http://localhost:${config.serverApi.port}`;

export default config;
