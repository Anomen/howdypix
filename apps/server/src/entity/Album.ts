import { Column, ViewEntity } from "typeorm";

@ViewEntity({
  expression: `
SELECT
   P.DIR,
   P.SOURCE,
   P.PARENTDIR,
   COUNT(DISTINCT P2.ID) AS NBPHOTOS,
   COUNT(DISTINCT P3.DIR) AS NBALBUMS,
   (
      SELECT
         FILE 
      FROM
         PHOTO P4 
      WHERE
         P4.DIR = P.DIR 
         AND P4.SOURCE = P.SOURCE LIMIT 0,
         1
   )
   AS PREVIEW 
FROM
   PHOTO P 
   LEFT JOIN
      PHOTO P2 
      ON (P2.DIR = P.DIR 
      AND P2.SOURCE = P.SOURCE) 
   LEFT JOIN
      PHOTO P3 
      ON (P3.PARENTDIR = P.DIR 
      AND P3.SOURCE = P.SOURCE) 
GROUP BY
   P.DIR,
   P.SOURCE,
   P.PARENTDIR;
    `,
})
export class Album {
  @Column("text")
  dir: string;

  @Column("text")
  source: string;

  @Column("text")
  parentDir: string;

  @Column("text")
  preview: string;

  @Column("int")
  nbPhotos: number;

  @Column("int")
  nbAlbums: number;
}
