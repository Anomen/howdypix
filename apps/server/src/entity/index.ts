export * from "./Source";
export * from "./Album";
export * from "./Photo";
export * from "./Search";
export * from "./SearchResult";
