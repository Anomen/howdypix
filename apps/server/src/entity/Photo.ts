import {
  Column,
  Entity,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
  Unique,
} from "typeorm";
import { SearchResult } from "./SearchResult";
import { Album } from "./Album";

export enum PHOTO_STATUS {
  NOT_PROCESSED = "not_processed",
  PROCESSED = "processed",
}

@Entity()
@Unique(["album", "file"])
export class Photo {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("text")
  status: PHOTO_STATUS;

  @Column("int")
  inode: number;

  @Column("int")
  mtime: number;

  @Column("int")
  ctime: number;

  @Column("int")
  birthtime: number;

  @Column("int")
  size: number;

  @Column("text", { nullable: true })
  make: string;

  @Column("text", { nullable: true })
  model: string;

  @Column("int", { nullable: true })
  ISO: number;

  @Column("int", { nullable: true })
  shutter: number;

  @Column("int", { nullable: true })
  processedShutter: number;

  @Column("int", { nullable: true })
  aperture: number;

  @Column("int", { nullable: true })
  processedAperture: number;

  @Column("int", { nullable: true })
  createDate: number;

  // TODO to remove
  @Column("text")
  source: string;

  // TODO to remove
  @Column("text")
  parentDir: string;

  // TODO to remove
  @Column("text")
  dir: string;

  @ManyToOne(() => Album, (album) => album.photos)
  public album!: Album;

  @Column("text")
  file: string;

  @OneToMany(() => SearchResult, (searchResult) => searchResult.photo)
  public searchResults!: SearchResult[];
}
