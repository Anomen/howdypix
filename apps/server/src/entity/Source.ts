import {
  Column,
  Entity,
  getConnection,
  Not,
  OneToMany,
  PrimaryGeneratedColumn,
  Unique,
} from "typeorm";
import { Photo as EntityPhoto } from "./Photo";
import { Album } from "./Album";

@Entity()
@Unique(["source"])
export class Source {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("text")
  source: string;

  @Column("text")
  dir: string;

  @OneToMany(() => Album, (album) => album.sourceLk)
  public albums!: Album[];

  async getNbPhotos(): Promise<number> {
    const photoRepository = getConnection().getRepository(EntityPhoto);

    const { count } = await photoRepository
      .createQueryBuilder()
      .select("COUNT(distinct file)", "count")
      .where({
        source: this.source,
        dir: "",
        file: Not(""),
      })
      .getRawOne();

    return count;
  }

  async getNbAlbums(): Promise<number> {
    const photoRepository = getConnection().getRepository(EntityPhoto);

    const { count } = await photoRepository
      .createQueryBuilder()
      .select("COUNT(distinct dir)", "count")
      .where({
        source: this.source,
        dir: Not(""),
      })
      .getRawOne();

    return count;
  }

  async getPreview(): Promise<string> {
    const photoRepository = getConnection().getRepository(EntityPhoto);

    const data = await photoRepository
      .createQueryBuilder()
      .select("file", "preview")
      .where({
        source: this.source,
        file: Not(""),
      })
      .getRawOne();

    // TODO change for undefined picture
    return data?.preview || "TOTO.JPG";
  }

  static async upsert(sourceName: string, dir: string): Promise<void> {
    const sourceRepository = getConnection().getRepository(Source);
    const sourceToUpdate = await sourceRepository.findOne({
      source: sourceName,
    });

    if (sourceToUpdate) {
      // Update the value
      sourceToUpdate.dir = dir;
      await sourceRepository.save(sourceToUpdate);
    } else {
      const source = new Source();
      source.source = sourceName;
      source.dir = dir;
      await sourceRepository.save(source);
    }
  }
}
