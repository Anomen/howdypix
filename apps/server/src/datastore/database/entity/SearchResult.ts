import {
  Column,
  Entity,
  PrimaryGeneratedColumn,
  ManyToOne,
  CreateDateColumn,
  UpdateDateColumn,
} from "typeorm";
import { Search } from "./Search";
import { Photo } from "./Photo";

@Entity()
export class SearchResult {
  @PrimaryGeneratedColumn()
  id: number;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @Column("int")
  searchId!: number;

  @Column("int")
  photoId!: number;

  @Column("int")
  public order!: number;

  @ManyToOne(() => Search, (search) => search.searchResults)
  public search!: Search;

  @ManyToOne(() => Photo, (photo) => photo.searchResults)
  public photo!: Photo;
}
