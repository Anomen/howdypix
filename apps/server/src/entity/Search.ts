import { Column, Entity, PrimaryGeneratedColumn, OneToMany } from "typeorm";
import { SearchResult } from "./SearchResult";

@Entity()
export class Search {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ type: "text", nullable: true })
  orderBy: string;

  @Column({ type: "text", nullable: true })
  filterBy: string;

  @Column({ type: "text", nullable: true })
  album: string;

  @Column({ type: "text", nullable: true })
  source: string;

  @OneToMany(
    () => SearchResult,
    searchResult => searchResult.search
  )
  public searchResults!: SearchResult[];
}
