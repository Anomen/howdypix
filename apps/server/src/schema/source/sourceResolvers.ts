import {
  NexusGenArgTypes,
  NexusGenFieldTypes,
} from "@howdypix/graphql-schema/schema.d";
import { generateThumbnailUrls } from "@howdypix/utils";
import { UserConfig, appConfig } from "../../config";
import { ApolloContext } from "../../types.d";
import { Source as EntitySource } from "../../entity/Source";

export const getSourcesResolver = (
  photoDirs: UserConfig["photoDirs"]
) => async (
  root: {},
  args: NexusGenArgTypes["Query"]["getAlbum"],
  ctx: ApolloContext
): Promise<NexusGenFieldTypes["Query"]["getSources"]> => {
  const sourceRepository = ctx.connection.getRepository(EntitySource);
  const sources: { [source: string]: EntitySource | null } = {};

  (await sourceRepository.find()).forEach((source) => {
    sources[source.source] = source;
  });

  return Object.keys(photoDirs).map((name) => ({
    name,
    nbAlbums: sources[name]?.nbAlbums ?? 0,
    nbPhotos: sources[name]?.nbPhotos ?? 0,
    preview:
      sources[name]?.preview &&
      generateThumbnailUrls(appConfig.webapp.baseUrl, {
        file: sources[name]?.preview,
        dir: sources[name]?.dir,
        source: name,
      }).map((tn) => tn.url)[0],
  }));
};
