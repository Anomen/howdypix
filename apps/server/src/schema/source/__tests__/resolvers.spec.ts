import { Connection } from "typeorm";
import { getSourcesResolver } from "../sourceResolvers";
import { UserConfig } from "../../../config";
import { Source as EntitySource } from "../../../entity/Source";

describe("currentUserResolver", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  const photoDirs: UserConfig["photoDirs"] = {
    source1: "source1",
    source2: "source2",
    source3: "source3",
  };

  const resolver = getSourcesResolver(photoDirs);

  const connection = ({
    getRepository: () => ({
      find: (): EntitySource[] => [
        {
          source: "source1",
          nbAlbums: 1,
          nbPhotos: 10,
          preview: "preview1.png",
          dir: "dir1",
        },
        {
          source: "source2",
          nbAlbums: 2,
          nbPhotos: 20,
          preview: "preview2.png",
          dir: "dir2",
        },
      ],
    }),
  } as unknown) as Connection;

  test("should return the current list of sources", async () => {
    expect(
      await resolver({}, { source: "source1" }, { user: null, connection })
    ).toMatchSnapshot();
  });
});
