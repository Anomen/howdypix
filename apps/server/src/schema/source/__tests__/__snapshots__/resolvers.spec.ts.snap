// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`currentUserResolver should return the current list of sources 1`] = `
Array [
  Object {
    "name": "source1",
    "nbAlbums": 1,
    "nbPhotos": 10,
    "preview": "http://localhost:3000/static/@source1:dir1/preview1x200.jpg",
  },
  Object {
    "name": "source2",
    "nbAlbums": 2,
    "nbPhotos": 20,
    "preview": "http://localhost:3000/static/@source2:dir2/preview2x200.jpg",
  },
  Object {
    "name": "source3",
    "nbAlbums": 0,
    "nbPhotos": 0,
    "preview": undefined,
  },
]
`;
